### PR TITLE
Fix behaviour of divergence in while loop conditions

### LINKED ISSUE #50856: 'a: while break 'a{} results in illigal instruction
Having the following main function results in an illigal instruction on stable, beta and nighlty.
It should be noted that when `n` is not used after the loop, the illigal instruction does not occur
https://play.rust-lang.org/?gist=96a11fd41327de4bf4e2c371dd7c1660&version=stable&mode=debug

```Rust
fn main() {
    let mut n = 0; 
    'a: while {break 'a; true} {
        n += 1;
    }
    n += 2;
}
```

Interestingly putting the same while-loop in a function works fine:
https://play.rust-lang.org/?gist=ed60d2118b970be39e09bc8653a25345&version=stable&mode=debug

### PR DESCRIPTION
**Pull request by varkor**
This fixes `'a: while break 'a {};` being treated as diverging, by tracking break expressions in the same way as in `loop` expressions.

Fixes #50856.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3855,10 +3855,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
                   // cannot use break with a value from a while loop
                   coerce: None,
-                  may_break: true,
+                  may_break: false,  // Will get updated if/when we find a `break`.
               };
 
-              self.with_breakable_ctxt(expr.id, ctxt, || {
+              let (ctxt, ()) = self.with_breakable_ctxt(expr.id, ctxt, || {
                   self.check_expr_has_type_or_error(&cond, tcx.types.bool);
                   let cond_diverging = self.diverges.get();
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: So what's confusing me here: there is somehow an implicit `break` should the condition be false (this is why `may_break` used to be considered `true`), and I can't see where that is accounted for.

For example, does this code accept:

```rust
let x: ! = {
  while false { return; };
};
```

( Do we have a test for this? It annoys me that this question is so hard to answer =) )

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3867,6 +3867,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                   self.diverges.set(cond_diverging);
               });
 
+              if ctxt.may_break {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: It seems like not having this before was the obvious bug...

File: src/librustc_typeck/check/mod.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_typeck/check/mod.rs]: Traditionally we have considered `while` loops as *always* potentially breaking — but it seems like your change could allow us to handle cases like this one (for example)

```rust
'a: loop {
    let x: ! = {
        while (continue 'a) { };
    };
}
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by varkor** [comment]
I've added a couple more tests. The actual error messages look a little weird due to https://github.com/rust-lang/rust/issues/50085, but that's an unrelated problem.

File: src/librustc_typeck/check/mod.rs [hunk 3]
```diff
@@ -3853,6 +3853,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                   self.diverges.set(cond_diverging);
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs]: Ah, this is the bit I was missing before (which ignores the divergence of the body). Very good.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**